src/Functions/reverseDNSQuery.cpp

```diff
@@ -0,0 +1,128 @@
+#include <Columns/ColumnString.h>
+#include <Columns/ColumnArray.h>
+#include <Functions/FunctionFactory.h>
+#include <Functions/IFunction.h>
+#include <DataTypes/DataTypeString.h>
+#include <DataTypes/DataTypeArray.h>
+#include <Common/DNSResolver.h>
+#include <Poco/Net/IPAddress.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+}
+
+class ReverseDNSQuery : public IFunction
+{
+public:
+    static constexpr auto name = "reverseDNSQuery";
+
+    static FunctionPtr create(ContextPtr)
+    {
+        return std::make_shared<ReverseDNSQuery>();
+    }
+
+    String getName() const override
+    {
+        return name;
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t) const override
+    {
+        if (arguments.empty())
+        {
+            throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH, "Function " + String(name) + "requires at least one argument");
+        }
+
+        auto ip_address_input = arguments[0].column;
+
+        auto ip_address = Poco::Net::IPAddress(ip_address_input->getDataAt(0).toString());
+
+        auto ptr_records = DNSResolver::instance().reverseResolve(ip_address);
+
+        auto string_column = ColumnString::create();
+
+        auto offsets_column = ColumnArray::ColumnOffsets::create();
+
+        if (!ptr_records.empty())
+        {
+            for (const auto & ptr_record : ptr_records)
+            {
+                string_column->insertData(ptr_record.c_str(), ptr_record.size());
+            }
+
+            offsets_column->insert(string_column->size());
+        }
+        else
+        {
+            offsets_column->insertDefault();
+        }
+
+        return ColumnArray::create(std::move(string_column), std::move(offsets_column));
```

Comment:
```suggestion
        auto res_type = getReturnTypeImpl(arguments);
        if (ptr_records.empty())
            return res_type->createColumnConstWithDefaultValue(input_rows_count);
            
        Array res;
        for (const auto & ptr_record : ptr_records)
        {
            res.push_back(ptr_record);
        }

        return res_type->createColumnConst(input_rows_count, res);
```

Reply:
(I suppose it will fix the issue with duplicated rows)

Reply:
Btw, `select reverseDNSQuery('8.8.8.8') from numbers(10)` throws LOGICAL_ERROR

Reply:
I have just applied your suggestion and it seems to work just fine:

```
arthur :) select reverseDNSQuery('8.8.8.8') from numbers(10)

SELECT reverseDNSQuery('8.8.8.8')
FROM numbers(10)

Query id: 39909f8b-7135-4789-9703-8b27e163fba9

┌─reverseDNSQuery('8.8.8.8')─┐
│ ['dns.google']             │
│ ['dns.google']             │
│ ['dns.google']             │
│ ['dns.google']             │
│ ['dns.google']             │
│ ['dns.google']             │
│ ['dns.google']             │
│ ['dns.google']             │
│ ['dns.google']             │
│ ['dns.google']             │
└────────────────────────────┘

10 rows in set. Elapsed: 0.006 sec. 
```